File: src/harness/schema.py
from __future__ import annotations

import json
from typing import Any, Dict

from .types import LLMAction


class SchemaError(ValueError):
    """Raised when LLM output fails strict schema validation."""


def _ensure_dict(raw: Any) -> Dict[str, Any]:
    if isinstance(raw, dict):
        return raw
    if isinstance(raw, str):
        try:
            decoded = json.loads(raw)
        except json.JSONDecodeError as exc:
            raise SchemaError(f"LLM output is not valid JSON: {exc}") from exc
        if not isinstance(decoded, dict):
            raise SchemaError("Decoded JSON must be an object.")
        return decoded
    raise SchemaError(f"LLM output must be dict or JSON string, got: {type(raw).__name__}")


def parse_llm_action(raw: Any) -> LLMAction:
    payload = _ensure_dict(raw)
    action_type = payload.get("type")
    if action_type not in {"tool_call", "final_response"}:
        raise SchemaError("Field 'type' must be 'tool_call' or 'final_response'.")

    if action_type == "final_response":
        content = payload.get("content")
        if not isinstance(content, str) or not content.strip():
            raise SchemaError("final_response requires non-empty string 'content'.")
        return LLMAction(action_type="final_response", raw_output=raw, content=content)

    tool_name = payload.get("tool_name")
    arguments = payload.get("arguments")
    if not isinstance(tool_name, str) or not tool_name:
        raise SchemaError("tool_call requires non-empty string 'tool_name'.")
    if not isinstance(arguments, dict):
        raise SchemaError("tool_call requires object 'arguments'.")

    return LLMAction(
        action_type="tool_call",
        raw_output=raw,
        tool_name=tool_name,
        arguments=arguments,
    )

File: ts/src/schema.ts
/**
 * LLM output parsing and strict schema validation.
 * All LLM output must pass through parseLLMAction() before use.
 */

import { SchemaError, type LLMAction } from "./types.js";

/**
 * Normalize raw input to a plain object.
 * Accepts objects or JSON strings, rejects everything else.
 */
export function ensureDict(raw: unknown): Record<string, unknown> {
  if (raw !== null && typeof raw === "object" && !Array.isArray(raw)) {
    return raw as Record<string, unknown>;
  }

  if (typeof raw === "string") {
    let decoded: unknown;
    try {
      decoded = JSON.parse(raw);
    } catch (exc) {
      throw new SchemaError(`LLM output is not valid JSON: ${(exc as Error).message}`);
    }
    if (decoded === null || typeof decoded !== "object" || Array.isArray(decoded)) {
      throw new SchemaError("Decoded JSON must be an object.");
    }
    return decoded as Record<string, unknown>;
  }

  const typeName = raw === null ? "null" : typeof raw;
  throw new SchemaError(`LLM output must be dict or JSON string, got: ${typeName}`);
}

/**
 * Parse and validate raw LLM output into a structured LLMAction.
 *
 * Validates:
 * - Input is a dict (or valid JSON string)
 * - `type` field is "tool_call" or "final_response"
 * - tool_call: `tool_name` is non-empty string, `arguments` is object
 * - final_response: `content` is non-empty string (after trim)
 */
export function parseLLMAction(raw: unknown): LLMAction {
  const payload = ensureDict(raw);

  const actionType = payload["type"];
  if (actionType !== "tool_call" && actionType !== "final_response") {
    throw new SchemaError("Field 'type' must be 'tool_call' or 'final_response'.");
  }

  if (actionType === "final_response") {
    const content = payload["content"];
    if (typeof content !== "string" || content.trim() === "") {
      throw new SchemaError("final_response requires non-empty string 'content'.");
    }
    return {
      action_type: "final_response",
      raw_output: raw,
      tool_name: null,
      arguments: {},
      content,
    };
  }

  // actionType === "tool_call"
  const toolName = payload["tool_name"];
  if (typeof toolName !== "string" || toolName === "") {
    throw new SchemaError("tool_call requires non-empty string 'tool_name'.");
  }

  const args = payload["arguments"];
  if (args === null || typeof args !== "object" || Array.isArray(args)) {
    throw new SchemaError("tool_call requires object 'arguments'.");
  }

  return {
    action_type: "tool_call",
    raw_output: raw,
    tool_name: toolName,
    arguments: args as Record<string, unknown>,
    content: null,
  };
}
